jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/metadata/accessors/objects/MetadataAnnotatedElement.java

```diff
@@ -202,17 +204,32 @@ public MetadataAnnotation getAnnotation(Class annotation) {
     /**
      * INTERNAL:
      * Return the annotated element for this accessor. Note: This method does
-     * not check against a metadata complete.
+     * not check against a meta-data complete.
+     * @param annotation annotation name
      */
-    public MetadataAnnotation getAnnotation(String annotation) {
+    public MetadataAnnotation getAnnotation(final String annotation) {
+        return getAnnotation(annotation, new HashSet<>());
+    }
+
+    /**
+     * INTERNAL:
+     * Return the annotated element for this accessor. Note: This method does
+     * not check against a meta-data complete.
+     * @param annotation annotation name
+     * @param names meta-annotations names cycle detection set, shall not be {@code null}
+     */
+    protected final MetadataAnnotation getAnnotation(final String annotation, final Set<String> names) {
         if (m_annotations == null && m_metaAnnotations == null) {
             return null;
         }
-
         MetadataAnnotation metadataAnnotation = m_annotations.get(annotation);
         if (metadataAnnotation == null) {
             for (MetadataAnnotation a: m_metaAnnotations.values()) {
-                MetadataAnnotation ma = m_factory.getMetadataClass(a.getName()).getAnnotation(annotation);
+                if (names.contains(a.getName()) && annotation.equals(a.getName())) {
```

Comment:
@Tomas-Kraus I think it still won't work. Could you add a test case where you invoke getAnnotation with an annotation name that is not present in the model for example "javax.persistence.Access", the system will loop recursively again because it won't match this "if" test. This "if" is only matched if the annotation is present. I would rewrite it with something like: 
`
if(annotation.equals(a.getName())return a; //existent recursive
if(names.contains(a.getName())continue; //non-existent recursive checked in past invocations
`

Reply:
This is 2.7 backport, let's resolve it in master first.